### PR TITLE
Add more typing to utils.events.containers

### PR DESCRIPTION
**Pull request by dstansby**
# References and relevant issues

# Description
Fixes three of the files completely, and handles all the other low hanging fruit in the other files in `napari.utils.events.containers.*`.

### CODE REVIEW

File: napari/utils/events/containers/_evented_list.py [hunk 1]
```diff
@@ -127,7 +128,7 @@ def __init__(
     # def extend(self, value: Iterable[_T]): ...
     # def remove(self, value: T): ...
 
-    def __setitem__(self, key, value):
+    def __setitem__(self, key, value: _T) -> None:
```
Comment (psobolewskiPhD) [inline napari/utils/events/containers/_evented_list.py:131]: Should `key` be typed as well? I think `key: Index`

File: napari/utils/events/containers/_selection.py [hunk 2]
```diff
@@ -124,21 +139,21 @@ def clear(self, keep_current: bool = False) -> None:
...
         self.add(obj)
 
     @classmethod
-    def __get_validators__(cls):
+    def __get_validators__(cls) -> Generator:
         yield cls.validate
 
     @classmethod
-    def validate(cls, v, field: 'ModelField'):
+    def validate(cls, v: Union["Selection", Dict], field: 'ModelField') -> "Selection":  # type: ignore[override]
```
Comment (psobolewskiPhD) [inline napari/utils/events/containers/_selection.py]: `v` could also be `sequence_like`, see line 170.
So should this `Union` also have `Sequence` ?
Looking a bit more, that pydantic check actually permits more then just Sequence, see:
https://github.com/pydantic/pydantic/blob/8364920acb272e2df5c6e261ca67622f3d6d1874/pydantic/v1/utils.py#L157-L158

**Review by psobolewskiPhD — APPROVED** [review]
Left two questions/comments, but everything else is good to me.

File: napari/utils/events/containers/_evented_list.py (hunk 1 above)
Comment (dstansby) [reply napari/utils/events/containers/_evented_list.py:131]: I wasn't sure what to type it as, so left it for now.

File: napari/utils/events/containers/_selection.py (hunk 2 above)
Comment (Czaki) [inline napari/utils/events/containers/_selection.py]:
```suggestion
    def validate(cls, v: Union['Selection', Dict], field: 'ModelField') -> 'Selection':  # type: ignore[override]
```

Comment (dstansby) [reply napari/utils/events/containers/_selection.py]: Does the type of quotes really matter here? And if it does, wouldn't it be better to enforce it with an automated pre-commit check?

File: napari/utils/events/containers/_evented_list.py [hunk 3]
```diff
@@ -304,7 +305,9 @@ def move_multiple(
         self.events.reordered(value=self)
         return len(move_plan)
 
-    def _move_plan(self, sources: Iterable[Index], dest_index: int):
+    def _move_plan(
+        self, sources: Iterable[Index], dest_index: int
+    ) -> Generator:
```
Comment (Czaki) [inline napari/utils/events/containers/_evented_list.py]:
```suggestion
    ) -> Generator[Tuple[int, int], None, None]:
```

File: napari/utils/events/containers/_nested_list.py [hunk 4]
```diff
@@ -205,7 +206,7 @@ def __setitem__(
         self, key: slice, value: Iterable[_T]
     ): ...  # pragma: no cover
 
-    def __setitem__(self, key: MaybeNestedIndex, value):
+    def __setitem__(self, key: MaybeNestedIndex, value) -> None:
```
Comment (Czaki) [inline napari/utils/events/containers/_nested_list.py]: why value do not have type annotation if overloaded version exists?

File: napari/utils/events/containers/_nested_list.py [hunk 5]
```diff
@@ -478,7 +479,9 @@ def _type_check(self, e) -> _T:
                 )
         return e
 
-    def _iter_indices(self, start=0, stop=None, root=()):
+    def _iter_indices(
+        self, start: int = 0, stop: Optional[int] = None, root=()
+    ):
```
Comment (Czaki) [inline napari/utils/events/containers/_nested_list.py]:
```suggestion
    def _iter_indices(
        self, start: int = 0, stop: Optional[int] = None, root: Tuple[int, ...]=()
    ) -> Generate[Union[int, Tuple[int]]]:
```

**Review by Czaki — CHANGES_REQUESTED** [review]

File: napari/utils/events/containers/_nested_list.py (hunk 4 above)
Comment (dstansby) [reply napari/utils/events/containers/_nested_list.py]: I thought overloaded implementations were meant to have no typing (https://docs.python.org/3/library/typing.html#overload), so I've removed the typing from it entirely and finished the typing in the overloaded signatures above.

File: napari/utils/events/containers/_nested_list.py (hunk 5 above)
Comment (dstansby) [reply napari/utils/events/containers/_nested_list.py]: 👍  done

File: napari/utils/events/containers/_evented_list.py (hunk 3 above)
Comment (dstansby) [reply napari/utils/events/containers/_evented_list.py]: 👍 done

**Review by Czaki — APPROVED** [review]

**Pull request merged**